tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -14212,6 +14250,86 @@ bool Sema::DiagnoseHLSLDecl(Declarator &D, DeclContext *DC, Expr *BitWidth,
         pMaxDispatchGrid = pAttr;
       }
       break;
+    case AttributeList::AT_HLSLNumThreads: {
+      const auto *SM =
+          hlsl::ShaderModel::GetByName(getLangOpts().HLSLProfile.c_str());
+
+      const std::string &entryName = getLangOpts().HLSLEntryFunction;
+
+      bool isEntry = false;
+      bool isNode = false;
+      bool isCS = false;
+      bool isMS = false;
+      bool isAS = false;
+
+      isEntry = !getLangOpts().IsHLSLLibrary &&
+                DC->getDeclKind() == Decl::Kind::TranslationUnit &&
+                functionName == entryName;
+      AttributeList *pAL = D.getDeclSpec().getAttributes().getList();
+      while (pAL) {
+        if (pAL->getKind() == AttributeList::AT_HLSLShader) {
+          Expr *ArgExpr = pAL->getArgAsExpr(0);
+          StringLiteral *Literal =
+              dyn_cast<StringLiteral>(ArgExpr->IgnoreParenCasts());
+          DXIL::ShaderKind Stage =
+              ShaderModel::KindFromFullName(Literal->getString());
+          isNode |= Stage == DXIL::ShaderKind::Node;
+          isCS |= Stage == DXIL::ShaderKind::Compute;
+          isMS |= Stage == DXIL::ShaderKind::Mesh;
+          isAS |= Stage == DXIL::ShaderKind::Amplification;          
+        }
+        pAL = pAL->getNext();
+      }           
+
+      if (!isNode && ((isEntry && !SM->IsCS() && !SM->IsMS() && !SM->IsAS()) ||
+                      (SM->IsLib() && !isCS && !isMS && !isAS))) {
+        Diag(pAttr->getLoc(), diag::err_hlsl_numthreads_attr);
+        result = false;
+      }
+      break;
+    }
+    case AttributeList::AT_HLSLWaveSize: {      
+      const auto *SM =
+          hlsl::ShaderModel::GetByName(getLangOpts().HLSLProfile.c_str());
+
+      const std::string &entryName = getLangOpts().HLSLEntryFunction;
+
+      bool isEntry = false;
+      bool isNode = false;
+      bool isCS = false;
+
+      isEntry = !SM->IsLib() &&
+                DC->getDeclKind() ==
+                    Decl::Kind::TranslationUnit &&
+                functionName == entryName;
+      
+      AttributeList *pAL = D.getDeclSpec().getAttributes().getList();
+      while (pAL) {
+        if (pAL->getKind() == AttributeList::AT_HLSLShader) {
+          Expr *ArgExpr = pAL->getArgAsExpr(0);
+          StringLiteral *Literal =
+              dyn_cast<StringLiteral>(ArgExpr->IgnoreParenCasts());
+          DXIL::ShaderKind Stage =
+              ShaderModel::KindFromFullName(Literal->getString());
+          isNode |= Stage == DXIL::ShaderKind::Node;
+          isCS |= Stage == DXIL::ShaderKind::Compute;          
+        }
+        pAL = pAL->getNext();
+      }
+      
+      if (!SM->IsLib() && !isEntry && !isNode) {
```

Comment:
In fact, without `!isCS`, this even fails if you put `[shader("compute")]` on that function.

Reply:
We need more tests for entry attributes where there is more than one entry function in the file, including when they don't have shader attributes, and when they are not the current entry for a non-lib-target compilation.

Reply:
I think the error is expected, the error that is emitted is:
"error: WaveSize attribute only valid on entry point functions"
The error is correct, because foo_wavesize as I understand is NOT an entry point function. If the error shouldn't be emitted, then the error message should be changed. Isn't it true that wave size attributes are only valid on entry point functions?

Reply:
I should also note that changing your example from `void main() {}` to `void main() {foo_wavesize();}` emits the same error even without the changes in this PR:
"error: WaveSize attribute only valid on entry point functions"
Your example only compiles successfully in the current version of the compiler (without the changes in my PR) because foo_wavesize() is unreachable from the entry point function, and so there is no codegen pass executed for the foo_wavesize function. 
Given this behavior and the fact that the diagnostic is moved to Sema, I think we can expect the error to be emitted given the example you gave above.

Reply:
This error isn't expected, unless we make a breaking change to HLSL language.  This is a commonly used pattern for HLSL to have multiple entry points in the same file and compile each separately.

In order to avoid that breaking change, we will need to follow control flow graph from the entry (or each entry and export for lib target), and collect all the functions that are used, then validate these.  This means this validation would need to happen under `DiagnoseTranslationUnit` instead of when each function is constructed.

If you call the function, however, we do need to validate the attributes as a called function, and in this case, they are not valid for a non-entry point function, so we would emit an error in that scenario.